ddtelemetry/src/fork.rs

```diff
@@ -0,0 +1,185 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present Datadog, Inc.
+use std::os::unix::net::{UnixDatagram, UnixStream};
+
+pub enum Fork {
+    Parent(libc::pid_t),
+    Child,
+}
+
+pub trait ForkSafe {}
+
+/// Forkable is meant to hold instances considered ForkSafe
+///
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct Forkable<T: ?Sized + 'static> {
+    inner: T,
+}
+
+impl<T: ?Sized> ForkSafe for Forkable<T> {}
+
+impl<T> Forkable<T> {
+    pub fn mark_as(inner: T) -> Self {
+        Self { inner }
+    }
+    pub fn take(self) -> T {
+        self.inner
+    }
+}
+
+impl<T: ?Sized> std::ops::Deref for Forkable<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+macro_rules! fork_safe {
+    ($($I:ident),*) => {
+        $(
+            impl ForkSafe for $I {}
+        )*
+    }
+}
+
+// Mark types known to be fork safe
+fork_safe!(UnixDatagram, UnixStream);
+fork_safe!(u8, u16, u32, u64, u128, i8, i16, i32, i64);
+
+macro_rules! impl_forksafe_for_tuple {
+    () => { impl ForkSafe for () {} };
+    ($A:ident $($I:ident)*) => {
+        impl_forksafe_for_tuple!($($I)*);
+
+        impl<$A : ForkSafe, $($I: ForkSafe),*> ForkSafe for ($A, $($I),*) {}
+    }
+}
+// Implement tuples "auto_trait" ForkSafe for *all tuples that contain only ForkSafe valuess
+impl_forksafe_for_tuple!(A B C D E F G H I J K L M);
+
+/// Forks process into a new standalone process
+///
+/// # Errors
+///
+/// This function will return an error if child process can't be forked
+///
+/// # Safety
+///
+/// Existing state of the process must allow safe forking, e.g. no background threads should be running
+/// as any locks held by these threads will be locked forever
+///
+pub unsafe fn fork() -> anyhow::Result<Fork> {
+    let res = libc::fork();
+    match res {
+        -1 => Err(anyhow::format_err!("Can't fork child process")),
+        0 => Ok(Fork::Child),
+        res => Ok(Fork::Parent(res)),
+    }
+}
+
+/// Forks supplied funciton pointer into a new process returning Pid of the process and call exit(0) once child function pointers
+/// stops executing
+///
+/// Function pointer is used to prevent capturing values in a closure, and all Args must be marked as ForkSafe
+///
+/// Marker ForkSafe should only be applied to types and traits that do not rely on global state
+/// especially Mutexes shared with already running threads must be avoided
+///
+/// # Errors
+/// function will return an Error if
+///
+/// This function will return an error if .
+pub fn safer_fork<Args>(args: Args, f: fn(Args) -> ()) -> anyhow::Result<libc::pid_t>
```

Comment:
I don't quite get why this is called "safer_fork", perhaps the documentation could be expanded to clarify it?

Reply:
If I had to guess, it would be because:
* `f` is a function pointer (not a closure) so it cannot capture values that are not fork-safe
* `Args` is fork-safe

So the only values the function called after fork"should be" able access are forksafe.

In practice I don't know if this is true. Globals for instance can still be accessed and be not fork safe.

For instance
```rust
lazy_static! {
    static ref ARRAY: Mutex<Vec<u8>> = Mutex::new(vec![]);
}
```

The Mutex could cause a deadlock if the fork tries to access it.

Reply:
Exactly, also a reason why Global mutexes are bad :tm:

Reply:
Exactly, also a reason why Global mutexes are bad :tm: